File: api/cloudflare_worker.js
/**
 * AI Daily Collector - Cloudflare Worker (完整版)
 * 支持: /api/hotspots, /api/stats, /api/sources, /rss
 */

// Primary data source for hotspots (legacy master branch)
const DATA_URL = "https://raw.githubusercontent.com/xxl115/ai-daily-collector/master/data/daily.json";
// Fallback data sources (try alternative branches in case of rename)
const FALLBACK_DATA_URLS = [
  DATA_URL,
  "https://raw.githubusercontent.com/xxl115/ai-daily-collector/main/data/daily.json",
];

export default {
  async fetch(request) {
    const url = new URL(request.url);
    const path = url.pathname;
    
    // CORS headers
    const corsHeaders = {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, OPTIONS',
      'Content-Type': 'application/json',
    };

    // Handle OPTIONS preflight
    if (request.method === 'OPTIONS') {
      return new Response(null, { status: 204, headers: corsHeaders });
    }

    try {
      // Root path - return info
      if (path === '/' || path === '') {
        return new Response(JSON.stringify({
          name: "AI Daily Collector",
          status: "ok",
          version: "1.1.0",
          endpoints: ["/", "/health", "/api/hotspots", "/api/stats", "/api/sources", "/rss"],
          documentation: "https://github.com/xxl115/ai-daily-collector"
        }), { headers: corsHeaders });
      }

      // Health check
      if (path === '/health') {
        return new Response(JSON.stringify({
          status: "ok",
          timestamp: new Date().toISOString(),
          version: "1.1.0"
        }), { headers: corsHeaders });
      }

      // Hotspots API - 获取热点文章
      if (path.startsWith('/api/hotspots')) {
        const limit = parseInt(url.searchParams.get('limit') || '50');
        const source = url.searchParams.get('source');  // 筛选来源
        const keyword = url.searchParams.get('keyword'); // 关键词搜索
        
      // Try multiple data sources to improve reliability
      let data = null;
      for (const url of FALLBACK_DATA_URLS) {
        try {
          const dataResp = await fetch(url);
          if (dataResp && dataResp.ok) {
            data = await dataResp.json();
            break;
          }
        } catch (e) {
          // ignore and try next source
        }
      }
      if (!data) {
        // As a last resort, provide an empty payload to avoid breaking the API.
        data = { hotspots: [], generated_at: new Date().toISOString(), total_collected: 0 };
      }
        let items = data.hotspots || [];
        
        // 来源筛选
        if (source && source !== 'all') {
          items = items.filter(item => item.source === source);
        }
        
        // 关键词搜索
        if (keyword) {
          const kw = keyword.toLowerCase();
          items = items.filter(item => 
            (item.title && item.title.toLowerCase().includes(kw)) ||
            (item.source && item.source.toLowerCase().includes(kw))
          );
        }
        
        // 限制数量并按热度排序
        items = items
          .sort((a, b) => (b.hot_score || 0) - (a.hot_score || 0))
          .slice(0, limit);
        
        // 统计来源
        const sourceCounts = {};
        (data.hotspots || []).forEach(item => {
          const s = item.source || 'Unknown';
          sourceCounts[s] = (sourceCounts[s] || 0) + 1;
        });
        
        return new Response(JSON.stringify({
          success: true,
          total: items.length,
          filtered: items.length,
          filters: {
            source: source || 'all',
            keyword: keyword || ''
          },
          sources: Object.entries(sourceCounts).map(([name, count]) => ({ name, count })),
          generated_at: data.generated_at || new Date().toISOString(),
          data: items
        }), { headers: corsHeaders });
      }

      // Stats API - 获取统计信息
      if (path === '/api/stats') {
        const dataResp = await fetch(DATA_URL);
        
        if (!dataResp.ok) {
          return new Response(JSON.stringify({
            success: false,
            error: "Failed to fetch data"
          }), { headers: corsHeaders });
        }
        
        const data = await dataResp.json();
        const items = data.hotspots || [];
        
        // 计算统计
        const total = items.length;
        const sources = new Set(items.map(i => i.source));
        const avgScore = total > 0 
          ? (items.reduce((sum, i) => sum + (i.hot_score || 0), 0) / total).toFixed(1)
          : 0;
        
        return new Response(JSON.stringify({
          success: true,
          today: {
            date: new Date().toISOString().split('T')[0],
            articles: total,
            hot_score_avg: parseFloat(avgScore)
          },
          total: {
            articles: total,
            sources: sources.size
          },
          generated_at: new Date().toISOString()
        }), { headers: corsHeaders });
      }

      // Sources API - 获取来源列表
      if (path === '/api/sources') {
        const dataResp = await fetch(DATA_URL);
        
        if (!dataResp.ok) {
          return new Response(JSON.stringify({
            success: false,
            error: "Failed to fetch data"
          }), { headers: corsHeaders });
        }
        
        const data = await dataResp.json();
        const items = data.hotspots || [];
        
        // 统计来源
        const sourceCounts = {};
        items.forEach(item => {
          const s = item.source || 'Unknown';
          sourceCounts[s] = (sourceCounts[s] || 0) + 1;
        });
        
        const sources = Object.entries(sourceCounts)
          .map(([name, count]) => ({ name, count }))
          .sort((a, b) => b.count - a.count);
        
        return new Response(JSON.stringify({
          success: true,
          total: sources.length,
          sources: sources
        }), { headers: corsHeaders });
      }

      // RSS feed
      if (path.startsWith('/rss')) {
        const limit = parseInt(url.searchParams.get('limit') || '20');
        const dataResp = await fetch(DATA_URL);
        
        if (!dataResp.ok) {
          return new Response('RSS unavailable', { status: 503 });
        }
        
        const data = await dataResp.json();
        const items = (data.hotspots || []).slice(0, limit);
        
        const xml = `<?xml version="1.0" encoding="UTF-8"?>
<rss version="2.0">
  <channel>
    <title>AI Daily - ${new Date().toLocaleDateString('zh-CN')}</title>
    <link>https://ai-daily-collector.workers.dev</link>
    <description>AI 热点资讯自动采集</description>
    <language>zh-CN</language>
    <lastBuildDate>${new Date().toUTCString()}</lastBuildDate>
    ${items.map(item => `
    <item>
      <title><![CDATA[${item.title || 'Untitled'}]]></title>
      <link>${item.url || '#'}</link>
      <description><![CDATA[来源: ${item.source || ''} | 热度: ${item.hot_score || 0}]]></description>
      <category>${item.source || 'AI'}</category>
      <pubDate>${item.published_at ? new Date(item.published_at).toUTCString() : new Date().toUTCString()}</pubDate>
    </item>`).join('')}
  </channel>
</rss>`;
        
        return new Response(xml, {
          headers: { 'Content-Type': 'application/rss+xml', ...corsHeaders }
        });
      }

      // 404 Not Found
      return new Response(JSON.stringify({ 
        error: "Not Found", 
        path: path,
        availableRoutes: ["/", "/health", "/api/hotspots", "/api/stats", "/api/sources", "/rss"]
      }), {
        status: 404,
        headers: corsHeaders
      });

    } catch (err) {
      return new Response(JSON.stringify({ 
        error: err.message,
        stack: process.env.DEBUG ? err.stack : undefined
      }), {
        status: 500,
        headers: corsHeaders
      });
    }
  },
};

File: api/cloudflare_worker.py
# -*- coding: utf-8 -*-
"""
Cloudflare Worker for AI Daily Collector

这个 Worker 充当 API 网关，提供:
- 缓存响应
- 请求限流
- API 路由转发

注意: Cloudflare Workers 主要支持 JavaScript
此文件用于参考，实际部署需要使用 wrangler 部署 JS 版本
"""

# 提示: 请使用 cloudflare_worker.js 作为实际 Worker 入口

WORKER_CODE = '''
// cloudflare_worker.js - Cloudflare Workers 入口
// 复制此代码到 Cloudflare Dashboard 或使用 wrangler 部署

export default {
  async fetch(request, env, ctx) {
    const url = new URL(request.url);
    const path = url.pathname;
    const method = request.method;

    // CORS 头
    const corsHeaders = {
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type',
    };

    // 处理 OPTIONS 请求
    if (method === 'OPTIONS') {
      return new Response(null, { headers: corsHeaders });
    }

    // API 路由
    try {
      // 健康检查
      if (path === '/health') {
        return new Response(JSON.stringify({
          status: 'ok',
          service: 'ai-daily-collector',
          timestamp: new Date().toISOString(),
        }), {
          headers: { ...corsHeaders, 'Content-Type': 'application/json' }
        });
      }

      // 热点列表
      if (path === '/api/hotspots' && method === 'GET') {
        // 检查缓存
        const cacheKey = `hotspots:${url.searchParams.get('limit') || 20}`;
        const cache = await env.CACHE.get(cacheKey);
        
        if (cache) {
          return new Response(cache, {
            headers: { ...corsHeaders, 'Content-Type': 'application/json', 'X-Cache': 'HIT' }
          });
        }

        // 转发到后端 API（需要配置你的后端地址）
        const backendUrl = `https://your-backend.workers.dev/api/hotspots${url.search}`;
        const response = await fetch(backendUrl, request);
        const data = await response.json();

        // 缓存响应（1小时）
        ctx.waitUntil(env.CACHE.put(cacheKey, JSON.stringify(data), { expirationTtl: 3600 }));

        return new Response(JSON.stringify(data), {
          headers: { ...corsHeaders, 'Content-Type': 'application/json', 'X-Cache': 'MISS' }
        });
      }

      // RSS Feed
      if (path === '/rss' || path === '/rss/latest') {
        const cacheKey = `rss:${path}`;
        const cache = await env.CACHE.get(cacheKey);
        
        if (cache) {
          return new Response(cache, {
            headers: { ...corsHeaders, 'Content-Type': 'application/rss+xml', 'X-Cache': 'HIT' }
          });
        }

        const backendUrl = `https://your-backend.workers.dev${path}`;
        const response = await fetch(backendUrl, request);
        const data = await response.text();

        ctx.waitUntil(env.CACHE.put(cacheKey, data, { expirationTtl: 1800 }));

        return new Response(data, {
          headers: { ...corsHeaders, 'Content-Type': 'application/rss+xml', 'X-Cache': 'MISS' }
        });
      }

      // 统计信息
      if (path === '/api/stats' && method === 'GET') {
        return new Response(JSON.stringify({
          sources: {
            rss: '10+ feeds',
            github: 'trending',
            hn: 'hacker news',
            ph: 'product hunt',
            newsnow: '20+ chinese platforms',
            v2ex: 'v2ex hot',
            reddit: '8+ subreddits',
          },
          features: {
            cache: 'Cloudflare KV',
            rate_limit: 'enabled',
            filter: 'keyword filtering',
            ranking: 'hotness scoring',
          },
          deploy: {
            docker: 'supported',
            workers: 'supported',
            k8s: 'supported',
          }
        }), {
          headers: { ...corsHeaders, 'Content-Type': 'application/json' }
        });
      }

      // 默认返回文档
      if (path === '/' || path === '/docs') {
        return new Response(`AI Daily Collector API

Endpoints:
- GET /health          - 健康检查
- GET /api/hotspots    - 热点列表
- GET /rss             - RSS Feed
- GET /rss/latest      - 最新 RSS
- GET /api/stats       - 统计信息

Deploy: https://github.com/xxl115/ai-daily-collector
`, {
          headers: { ...corsHeaders, 'Content-Type': 'text/plain' }
        });
      }

      // 404
      return new Response('Not Found', { status: 404, headers: corsHeaders });

    } catch (error) {
      return new Response(JSON.stringify({
        error: error.message,
        stack: error.stack,
      }), {
        status: 500,
        headers: { ...corsHeaders, 'Content-Type': 'application/json' }
      });
    }
  }
};
'''

# 使用说明
USAGE = '''
# Cloudflare Workers 部署

## 方式 1: 使用 Wrangler CLI

```bash
# 1. 安装 Wrangler
npm install -g wrangler

# 2. 登录 Cloudflare
wrangler login

# 3. 创建 KV 命名空间
wrangler kv:namespace create CACHE

# 4. 编辑 wrangler.toml，填入 KV ID

# 5. 部署
wrangler deploy

# 或者开发模式
wrangler dev
```

## 方式 2: Cloudflare Dashboard

1. 访问 [Cloudflare Dashboard](https://dash.cloudflare.com)
2. 进入 "Workers & Pages" → "Create Application"
3. 点击 "Create Worker"
4. 粘贴上面的 JavaScript 代码
5. 配置 KV 绑定
6. 点击 "Deploy"

## 方式 3: GitHub Actions

在 deploy.yml 中添加:

```yaml
deploy-cloudflare:
  runs-on: ubuntu-latest
  steps:
    - uses: actions/checkout@v4
    
    - name: Deploy to Cloudflare Workers
      uses: cloudflare/wrangler-action@v3
      with:
        apiToken: ${{ secrets.CF_API_TOKEN }}
        command: deploy
```

## 环境变量 Secrets

在 GitHub 仓库添加:
- CF_API_TOKEN: Cloudflare API Token
- CF_ACCOUNT_ID: Cloudflare Account ID
'''

if __name__ == "__main__":
    print("=" * 60)
    print("Cloudflare Worker for AI Daily Collector")
    print("=" * 60)
    print()
    print("注意: Cloudflare Workers 使用 JavaScript")
    print("请使用 cloudflare_worker.js 作为 Worker 入口")
    print()
    print(USAGE)
